lib/puppet/configurer.rb

```diff
@@ -225,12 +225,26 @@ def run(options = {})
         # mode. We shouldn't try to do any failover in that case.
         if options[:catalog].nil? && do_failover
           server, port = find_functional_server
-          if server.nil?
-            raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
-          else
-            #TRANSLATORS 'server_list' is the name of a setting and should not be translated
-            Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
-            report.master_used = "#{server}:#{port}"
+          begin
+            if server.nil?
+              raise Puppet::Error, _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            else
+              #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+              Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
+              report.master_used = "#{server}:#{port}"
+            end
+          rescue Puppet::Error => detail
+            if Puppet[:usecacheonfailure]
+              options[:pluginsync] = false
+              @running_failure = true
+              if server.nil?
+                server = Puppet[:server_list].first[0]
+                port = Puppet[:server_list].first[1] || Puppet[:masterport]
```

Comment:
Is there any possibility that `Puppet[:server_list]` is `nil` or `[]` here?

Reply:
I suppose not, since we are in block starting here:
https://github.com/puppetlabs/puppet/blob/139b5e93a2bfa82afd9c6aa1a04db5f3c6e8bd94/lib/puppet/configurer.rb#L226
and `do_failover` is set only if  `Puppet[:server_list]` has at least one element:
https://github.com/puppetlabs/puppet/blob/139b5e93a2bfa82afd9c6aa1a04db5f3c6e8bd94/lib/puppet/configurer.rb#L219-L223